tests/python/unittest/test_runtime_graph.py

```diff
@@ -68,5 +69,86 @@ def check_remote():
     check_verify()
     check_remote()
 
+def test_graph_symbolic_shape():
+    b1 = tvm.var("n1")
+    b2 = tvm.var("n2")
+    b3 = tvm.var("n3")
+
+    shape_var_bounds = {
+        b1 : 128,
+        b2 : 32,
+        b3 : 12
+    }
+
+    x = relay.var("x",
+              shape=[b1,
+                     4,
+                     2, 3], dtype="float32")
+
+
+
+
+    y = relay.var("y",
+              shape=[b2,
+                     4,
+                     2, 3], dtype="float32")
+
+    z = relay.op.tensor.concatenate([x, y], axis=0)
+
+
+    a = relay.var("a",
+              shape=[b3,
+                     4,
+                     2, 3], dtype="float32")
+    b = relay.var("b",
+              shape=[27,
+                     4,
+                     2, 3], dtype="float32")
+    c = relay.op.tensor.concatenate([a, b], axis=0)
+
+    out = relay.op.tensor.concatenate([z, c], axis=0)
+
+    func = relay.Function([x, y, a, b], out)
+    graph, mod, param = relay.build_module.build(func, target="llvm", shape_var_bounds=shape_var_bounds)
+    rt = tvm.contrib.graph_runtime.create(graph, mod, tvm.cpu())
+    #### Test
+    def test(n1, n2, n3):
+       import numpy as np
+       vars = {
+              "n1": n1,
+              "n2": n2,
+              "n3": n3
+       }
+       shapes = {
+              "x": (n1, 4, 2, 3),
+              "y": (n2, 4, 2, 3),
+              "a": (n3, 4, 2, 3),
+              "b": (27, 4, 2, 3)
+       }
+       arrays = {
+              "x": np.random.uniform(-1, 1, shapes["x"]).astype("float32"),
+              "y": np.random.uniform(-1, 1, shapes["y"]).astype("float32"),
+              "a": np.random.uniform(-1, 1, shapes["a"]).astype("float32"),
+              "b": np.random.uniform(-1, 1, shapes["b"]).astype("float32"),
+       }
+       inputs = {k:tvm.nd.array(v) for k, v in arrays.items()}
+ 
+       z = np.vstack((arrays["x"], arrays["y"]))
+       c = np.vstack((arrays["a"], arrays["b"]))
+       ans = np.vstack((z, c))
+       rt.set_shape_variable(vars)
```

Comment:
Is it possible to move this to the beginning of set_input? Maybe we can create a map from name of var to input tensor index and shape index, and generate var_bounds with input DLTensor shape. In this way user can have a uniform runtime interface for both static and dynamic shape.

Reply:
Yes. Will do it today

Reply:
@kevinthesun I don't think update from C++ side is a good solution because in C++ side we have to create correct view before copy. In order to add this sugar we may modify JSON format again.

Reply:
@kevinthesun Let's discuss this workflow. I think in order to add this sugar, we have two potential way. Both are not elegant. 

0. ```set_input``` workflow:
When we set inputs, we have to set one by one. Each one requires correct view to make copy successful, which means, if we update view after setting each single one input, we will generate a lot of useless view unless we have kind of way to force copy before change view.

1. Handle it in C++
Similar to 0, we need a way to brute copy when set input, and update view when all copies are finished.

2. Handle it in Python 
We have to know symbolic variable location, which is a map<input_node_name, <dim_location, variable_name> . In order to get this one, we can either obtain from C++ when parse the expression shape, or adding these info in JSON file, because in runtime module we can't get the json back to python any more. Dumping from C++ requires adding more FFI feature to tvm runtime pack function, which seems, useless now. If we add to JSON, we are changing the model definition once more.